### PR TITLE
[Static if] replace overload constraints with static if (sorting.d)

### PR DESCRIPTION
**Pull request by wilzbach**
Split up of #5145

### CODE REVIEW

File: std/algorithm/sorting.d [hunk 1]
```diff
@@ -3753,57 +3753,45 @@ void topNIndex(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable,
 if (isRandomAccessRange!Range &&
     isRandomAccessRange!RangeIndex &&
     hasAssignableElements!RangeIndex &&
-    isIntegral!(ElementType!(RangeIndex)))
+        isIntegral!(ElementType!(RangeIndex)) || is(ElementType!(RangeIndex) == ElementType!(Range)*))
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: The next, to be grouped overload is
 ```d
  isRandomAccessRange!Range &&
  isRandomAccessRange!RangeIndex &&    
  hasAssignableElements!RangeIndex &&
  is(ElementType!(RangeIndex) == ElementType!(Range)*))
```
Notice that the first three constraints are exactly equal and only the last one differs as the first overload is for a range of indexes and the second one for a range of pointers.

Comment (John-Colvin) [inline std/algorithm/sorting.d]: What's with the extra 4 leading spaces?

I suggest adding parens around this line to make the logic grouping clearer.

File: std/algorithm/sorting.d [hunk 2]
```diff
@@ -3753,57 +3753,45 @@ void topNIndex(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable,
...
 
-    if (index.empty) return;
-    static bool indirectLess(const ElementType!(RangeIndex) a,
-                             const ElementType!(RangeIndex) b)
-    {
-        return binaryFun!less(*a, *b);
     }
-    auto heap = BinaryHeap!(RangeIndex, indirectLess)(index, 0);
-    foreach (i; 0 .. r.length)
+    else
```
Comment (John-Colvin) [inline std/algorithm/sorting.d]: Might be good to actually have a `static if` here as well and an `else static assert(false)` at the end to help ensure that the constraint and `static if`s stay in sync.

**Comment by John-Colvin** [comment]
Other than the minor comments above, LGTM

**Comment by quickfur** [comment]
ping @wilzbach 

Let's address @John-Colvin 's comments and let's get this PR merged. It's been sitting here for far too long already.

File: std/algorithm/sorting.d [hunk 3]
```diff
@@ -3748,58 +3748,46 @@ void topNIndex(alias less = "a < b", SwapStrategy ss = SwapStrategy.unstable,
               (Range r, RangeIndex index, SortOutput sorted = No.sortOutput)
 if (isRandomAccessRange!Range &&
     isRandomAccessRange!RangeIndex &&
-    hasAssignableElements!RangeIndex &&
-    isIntegral!(ElementType!(RangeIndex)))
+    hasAssignableElements!RangeIndex)
```
Comment (wilzbach) [inline std/algorithm/sorting.d]: As suggested I moved the specific ElementType checks and trigger a static assert error (and thus nice error message!) if an invalid type is passed by accident.

**Comment by wilzbach** [comment]
> Let's address @John-Colvin 's comments and let's get this PR merged. It's been sitting here for far too long already.

I absolutely agree & I am quite sorry that I have been so busy over the last weeks.

**Comment by John-Colvin** [comment]
LGTM

**Comment by andralex** [comment]
@wilzbach you're on fire over there

**Comment by wilzbach** [comment]
> @wilzbach you're on fire over there

Isn't most of the Phobos queue on 🔥?

**Pull request merged**